ERROR: While parsing vendor/composer/composer/src/Composer/Downloader/DownloadManager.php

https://raw.githubusercontent.com/composer/composer/master/src/Composer/Downloader/DownloadManager.php

```
$ phan vendor/composer/composer/src/Composer/Downloader/DownloadManager.php
[info] Disabling Xdebug: Phan is around five times as slow when Xdebug is enabled (Xdebug only makes sense when debugging Phan itself)
[info] To run Phan with Xdebug, set the environment variable PHAN_ALLOW_XDEBUG to 1.
[info] To disable this warning, set the environment variable PHAN_DISABLE_XDEBUG_WARN to 1.
[info] To include function signatures of Xdebug, see .phan/internal_stubs/xdebug.phan_php
[debug] Checking PHAN_ALLOW_XDEBUG
[debug] Because xdebug was installed, Phan will restart.
[debug] The Xdebug extension is loaded (2.9.2)
[debug] Process restarting (PHAN_ALLOW_XDEBUG=internal|2.9.2|1|*|*)
[debug] Running /usr/bin/php7.4 -n -c /tmp/1KVmtN /usr/local/bin/phan vendor/composer/composer/src/Composer/Downloader/DownloadManager.php
ERROR: While parsing vendor/composer/composer/src/Composer/Downloader/DownloadManager.php...
AssertionError: Expected union type_parts to contain a single type in phar:///usr/local/bin/phan.phar/src/Phan/Language/Type/IntersectionType.php:141
Stack trace:
#0 phar:///usr/local/bin/phan.phar/src/Phan/Language/Type/IntersectionType.php(88): Phan\Language\Type\IntersectionType::flattenTypes()
#1 phar:///usr/local/bin/phan.phar/src/Phan/Language/Type.php(1288): Phan\Language\Type\IntersectionType::createFromTypes()
#2 phar:///usr/local/bin/phan.phar/src/Phan/Language/Type.php(822): Phan\Language\Type::parseIntersectionTypeFromTemplateParameterList()
#3 phar:///usr/local/bin/phan.phar/src/Phan/Language/Type.php(1516): Phan\Language\Type::fromInternalTypeName()
#4 phar:///usr/local/bin/phan.phar/src/Phan/Language/UnionType.php(445): Phan\Language\Type::fromStringInContext()
#5 phar:///usr/local/bin/phan.phar/src/Phan/Language/Element/Comment/Builder.php(267): Phan\Language\UnionType::fromStringInContext()
#6 phar:///usr/local/bin/phan.phar/src/Phan/Language/Element/Comment/Builder.php(684): Phan\Language\Element\Comment\Builder->returnTypeFromCommentLine()
#7 phar:///usr/local/bin/phan.phar/src/Phan/Language/Element/Comment/Builder.php(422): Phan\Language\Element\Comment\Builder->maybeParseReturn()
#8 phar:///usr/local/bin/phan.phar/src/Phan/Language/Element/Comment/Builder.php(328): Phan\Language\Element\Comment\Builder->parseCommentLine()
#9 phar:///usr/local/bin/phan.phar/src/Phan/Language/Element/Comment.php(421): Phan\Language\Element\Comment\Builder->build()
#10 phar:///usr/local/bin/phan.phar/src/Phan/Language/Element/Method.php(521): Phan\Language\Element\Comment::fromStringInContext()
#11 phar:///usr/local/bin/phan.phar/src/Phan/Parse/ParseVisitor.php(332): Phan\Language\Element\Method::fromNode()
#12 phar:///usr/local/bin/phan.phar/src/Phan/Analysis.php(190): Phan\Parse\ParseVisitor->visitMethod()
#13 phar:///usr/local/bin/phan.phar/src/Phan/Analysis.php(217): Phan\Analysis::parseNodeInContext()
#14 phar:///usr/local/bin/phan.phar/src/Phan/Analysis.php(217): Phan\Analysis::parseNodeInContext()
#15 phar:///usr/local/bin/phan.phar/src/Phan/Analysis.php(217): Phan\Analysis::parseNodeInContext()
#16 phar:///usr/local/bin/phan.phar/src/Phan/Analysis.php(143): Phan\Analysis::parseNodeInContext()
#17 phar:///usr/local/bin/phan.phar/src/Phan/Phan.php(294): Phan\Analysis::parseFile()
#18 phar:///usr/local/bin/phan.phar/src/phan.php(38): Phan\Phan::analyzeFileList()
#19 /usr/local/bin/phan.phar(6): require('phar:///usr/loc...')
#20 {main}
[debug] Restarted process exited 1
```

problem is with the weird phpdoc:
```
    /**
     * @return array<'dist'|'source'>&non-empty-array
     */
```

--- a/NEWS.md
+++ b/NEWS.md
@@ -26,6 +26,7 @@ Bug fixes:
 - Fix false positive `PhanPossiblyUndefinedGlobalVariable*` instance when `global $var` is used within a conditional. (#4539)
 - Fix false positive `PhanPluginRedundantAssignmentInLoop` instance when a variable is modified in a catch statement with a break/continue. (#4542)
 - Fix some incorrect line numbers in some plugin issues.
+- Fix crash seen when parsing intersection types containing union types such as `non-empty-array&array<'a'|'b'>` (#4544)
 
 Maintenance:
 - Fix old return type signature for `get_headers` (#3273)

--- a/internal/Issue-Types-Caught-by-Phan.md
+++ b/internal/Issue-Types-Caught-by-Phan.md
@@ -4958,6 +4958,12 @@ Saw an @param annotation for ${PARAMETER}, but it was not found in the param lis
 
 e.g. [this issue](https://github.com/phan/phan/tree/v5/tests/files/expected/0373_reject_bad_type_narrowing.php.expected#L1) is emitted when analyzing [this PHP file](https://github.com/phan/phan/tree/v5/tests/files/src/0373_reject_bad_type_narrowing.php#L4).
 
+## PhanCommentUnsupportedUnionType
+
+```
+Saw a union type {TYPE} with more than 1 type in a location that does not support union types
+```
+
 ## PhanCommentVarInsteadOfParam
 
 ```

--- a/src/Phan/Issue.php
+++ b/src/Phan/Issue.php
@@ -650,6 +650,7 @@ class Issue
     public const CommentDuplicateMagicMethod      = 'PhanCommentDuplicateMagicMethod';
     public const CommentDuplicateMagicProperty    = 'PhanCommentDuplicateMagicProperty';
     public const CommentObjectInClassConstantType = 'PhanCommentObjectInClassConstantType';
+    public const CommentUnsupportedUnionType      = 'PhanCommentUnsupportedUnionType';
     // phpcs:enable Generic.NamingConventions.UpperCaseConstantName.ClassConstantNotUpperCase
     // end of issue name constants
 
@@ -5585,6 +5586,14 @@ class Issue
                 "Impossible phpdoc declaration that a class constant {CONST} has a type {TYPE} containing objects. This type is ignored during analysis.",
                 self::REMEDIATION_B,
                 16021
+            ),
+            new Issue(
+                self::CommentUnsupportedUnionType,
+                self::CATEGORY_COMMENT,
+                self::SEVERITY_LOW,
+                "Saw a union type {TYPE} with more than 1 type in a location that does not support union types",
+                self::REMEDIATION_B,
+                16027
             ),
         ];
         // phpcs:enable Generic.Files.LineLength

--- a/src/Phan/Language/EmptyUnionType.php
+++ b/src/Phan/Language/EmptyUnionType.php
@@ -997,6 +997,15 @@ final class EmptyUnionType extends UnionType
 
     /**
      * @return bool
+     * True if this is non-empty and exclusively object types.
+     */
+    public function isObject(): bool
+    {
+        return false;  // empty
+    }
+
+    /**
+     * @return bool
      * True if any of the types in this UnionType made $matcher_callback return true
      */
     public function hasTypeMatchingCallback(Closure $matcher_callback): bool

--- a/src/Phan/Language/Type.php
+++ b/src/Phan/Language/Type.php
@@ -1283,7 +1283,7 @@ class Type implements Stringable
         ?CodeBase $code_base,
         ?Context $context
     ): Type {
-        $result = IntersectionType::createFromTypes($template_parameter_type_list, $code_base, $context);
+        $result = IntersectionType::createFromTypes($template_parameter_type_list, $code_base, $context, true);
         return $is_nullable ? $result->withIsNullable(true) : $result;
     }
 

--- a/src/Phan/Language/Type/CallableStringType.php
+++ b/src/Phan/Language/Type/CallableStringType.php
@@ -153,6 +153,6 @@ final class CallableStringType extends StringType implements CallableInterface
      */
     protected function isSubtypeOfNonNullableType(Type $type, CodeBase $code_base): bool
     {
-        return $type instanceof CallableType || \get_class($type) === StringType::class || $type instanceof MixedType;
+        return $type instanceof CallableType || \in_array(\get_class($type), [StringType::class, NonEmptyStringType::class], true) || $type instanceof MixedType;
     }
 }

--- a/src/Phan/Language/Type/IntersectionType.php
+++ b/src/Phan/Language/Type/IntersectionType.php
@@ -82,10 +82,11 @@ final class IntersectionType extends Type
      * @param non-empty-list<Type|UnionType> $types
      * @param ?CodeBase $code_base
      * @param ?Context $context @unused-param
+     * @param bool $suppress_union_type_warnings
      */
-    public static function createFromTypes(array $types, ?CodeBase $code_base, ?Context $context): Type
+    public static function createFromTypes(array $types, ?CodeBase $code_base, ?Context $context, bool $suppress_union_type_warnings = false): Type
     {
-        $new_types = self::flattenTypes($types);
+        $new_types = self::flattenTypes($types, $code_base, $context, $suppress_union_type_warnings);
         if (!$new_types) {
             throw new AssertionError('Saw empty list of type_parts in intersection type');
         }
@@ -131,14 +132,37 @@ final class IntersectionType extends Type
      * @param non-empty-list<Type|UnionType> $types
      * @return non-empty-list<Type>
      */
-    private static function flattenTypes(array $types): array
+    private static function flattenTypes(array $types, ?CodeBase $code_base, ?Context $context, bool $suppress_union_type_warnings): array
     {
         $new_types = [];
         foreach ($types as $type) {
             if ($type instanceof UnionType) {
                 $type_set = $type->getTypeSet();
                 if (count($type_set) !== 1) {
-                    throw new AssertionError("Expected union type_parts to contain a single type");
+                    if ($suppress_union_type_warnings) {
+                        if ($code_base && $context) {
+                            Issue::maybeEmit(
+                                $code_base,
+                                $context,
+                                Issue::CommentUnsupportedUnionType,
+                                $context->getLineNumberStart(),
+                                $type
+                            );
+                        }
+                        $type = $type->withFlattenedArrayShapeOrLiteralTypeInstances();
+                        $type_set = $type->getTypeSet();
+                        if (count($type_set) !== 1) {
+                            if ($type->isArray()) {
+                                $type_set = [ArrayType::instance($type->containsNullable())];
+                            } elseif ($type->isObject()) {
+                                $type_set = [ObjectType::instance($type->containsNullable())];
+                            } else {
+                                $type_set = [MixedType::instance(false)];
+                            }
+                        }
+                    } else {
+                        throw new AssertionError("Expected union type_parts to contain a single type");
+                    }
                 }
                 $type = reset($type_set);
                 if (!$type instanceof Type) {
@@ -152,7 +176,13 @@ final class IntersectionType extends Type
                 }
             }
         }
+        $has_exclusively_truthy = false;
+        $has_array = false;
+        $has_object = false;
         foreach ($new_types as $i => $type) {
+            if (!$has_exclusively_truthy) {
+                $has_exclusively_truthy = $type->isAlwaysTruthy();
+            }
             // Convert callable&object to callable-object, etc.
             if ($type instanceof CallableType) {
                 foreach ($new_types as $j => $other) {
@@ -171,7 +201,53 @@ final class IntersectionType extends Type
                     }
                 }
             }
+            if ($type instanceof ObjectType) {
+                $has_object = true;
+            } elseif ($type instanceof ArrayType) {
+                $has_array = true;
+            }
             // TODO: Convert iterable<k,v> to Traversable<k,v> if another value is an object
+        }
+        if ($has_object) {
+            foreach ($new_types as $j => $other) {
+                if ($other->isObject()) {
+                    // do nothing
+                } elseif (!$other->isPossiblyObject()) {
+                    // Emit PhanImpossibleIntersectionType elsewhere
+                    continue;
+                } else {
+                    $other = $other->asObjectType();
+                    if ($other) {
+                        $new_types[$j] = $other;
+                    }
+                }
+            }
+        } elseif ($has_array) {
+            foreach ($new_types as $j => $other) {
+                if ($other instanceof ArrayType) {
+                    continue;
+                }
+                if (!$other instanceof IterableType && !$other instanceof MixedType) {
+                    // Emit PhanImpossibleIntersectionType elsewhere
+                    continue;
+                }
+                $other = $other->asArrayType();
+                if ($other) {
+                    $new_types[$j] = $other;
+                }
+            }
+        }
+        if ($has_exclusively_truthy) {
+            foreach ($new_types as $j => $other) {
+                if ($other->isAlwaysTruthy()) {
+                    continue;
+                }
+                if (!$other->isPossiblyTruthy()) {
+                    // Emit PhanImpossibleIntersectionType elsewhere
+                    continue;
+                }
+                $new_types[$j] = $other->asNonFalseyType();
+            }
         }
         return \array_values($new_types);
     }

--- a/src/Phan/Language/UnionType.php
+++ b/src/Phan/Language/UnionType.php
@@ -3840,6 +3840,21 @@ class UnionType implements Serializable, Stringable
 
     /**
      * @return bool
+     * True if this is non-empty and exclusively object types
+     */
+    public function isObject(): bool
+    {
+        if ($this->isEmpty()) {
+            return false;
+        }
+
+        return $this->allTypesMatchCallback(static function (Type $type): bool {
+            return $type->isObject();
+        });
+    }
+
+    /**
+     * @return bool
      * True if any of the types in this UnionType made $matcher_callback return true
      */
     public function hasTypeMatchingCallback(Closure $matcher_callback): bool

--- a/tests/Phan/Language/UnionTypeTest.php
+++ b/tests/Phan/Language/UnionTypeTest.php
@@ -22,6 +22,7 @@ use Phan\Language\Type\IntersectionType;
 use Phan\Language\Type\IntType;
 use Phan\Language\Type\IterableType;
 use Phan\Language\Type\MixedType;
+use Phan\Language\Type\NonEmptyGenericArrayType;
 use Phan\Language\Type\ObjectType;
 use Phan\Language\Type\ResourceType;
 use Phan\Language\Type\StaticType;
@@ -300,12 +301,12 @@ final class UnionTypeTest extends BaseTest
     private static function makePHPDocUnionType(string $union_type_string): UnionType
     {
         self::assertTrue(\preg_match(self::VALID_UNION_TYPE_REGEX, $union_type_string) > 0, "$union_type_string should be parsed by regex");
-        return UnionType::fromStringInContext($union_type_string, new Context(), Type::FROM_PHPDOC);
+        return UnionType::fromStringInContext($union_type_string, new Context(), Type::FROM_PHPDOC, self::$code_base);
     }
 
     private static function makePHPDocType(string $type): Type
     {
-        return Type::fromStringInContext($type, new Context(), Type::FROM_PHPDOC);
+        return Type::fromStringInContext($type, new Context(), Type::FROM_PHPDOC, self::$code_base);
     }
 
     private function assertIsType(Type $type, string $union_type_string): void
@@ -818,6 +819,26 @@ final class UnionTypeTest extends BaseTest
         $this->assertSame(1, $type->typeCount());
         $this->assertSame('\MyClass&\MyInterfaceUTT', (string)$type);
         $this->assertInstanceOf(IntersectionType::class, $type->getTypeSet()[0]);
+    }
+
+    public function testIntersectionTypeArray(): void
+    {
+        $type = self::makePHPDocUnionType('non-empty-array&array<0|1>');
+        $this->assertSame(1, $type->typeCount());
+        // Currently, not able to use union types as a part of intersection types when building them.
+        $this->assertSame('non-empty-array<mixed,int>', (string)$type);
+        $this->assertInstanceOf(NonEmptyGenericArrayType::class, $type->getTypeSet()[0]);
+        $type = self::makePHPDocUnionType('non-empty-array&iterable<string,array>');
+        $this->assertSame('non-empty-array<string,array>', (string)$type);
+        $this->assertInstanceOf(NonEmptyGenericArrayType::class, $type->getTypeSet()[0]);
+        $type = self::makePHPDocUnionType('object&iterable<string,array>');
+        $this->assertSame('\Traversable<string,array>', (string)$type);
+        $this->assertInstanceOf(Type::class, $type->getTypeSet()[0]);
+        // Should not crash when given unsupported combinations of types.
+        // Currently, union types aren't allowed in intersection types.
+        $type = self::makePHPDocUnionType('array<string|int>&array<string|float>');
+        $this->assertSame('array', (string)$type);
+        $this->assertInstanceOf(ArrayType::class, $type->getTypeSet()[0]);
     }
 
     public function testIntersectionTypeWithUnionType(): void

--- a/tests/php80_files/expected/041_intersection_type_phpdoc.php.expected
+++ b/tests/php80_files/expected/041_intersection_type_phpdoc.php.expected
@@ -1,7 +1,7 @@
 %s:8 PhanDebugAnnotation @phan-debug-var requested for variable $a - it has union type callable-object
-%s:8 PhanDebugAnnotation @phan-debug-var requested for variable $b - it has union type callable-array&string[]
+%s:8 PhanDebugAnnotation @phan-debug-var requested for variable $b - it has union type callable-array&non-empty-array<mixed,string>
 %s:8 PhanDebugAnnotation @phan-debug-var requested for variable $c - it has union type callable-string
 %s:16 PhanUnreferencedPublicMethod Possibly zero references to public method \A41::method()
 %s:22 PhanTypeMismatchArgumentProbablyReal Argument 1 ($a) is [A41::class,'method'] of type array{0:'A41',1:'method'} but \test41() takes callable-object (no real type) defined at %s:8 (the inferred real argument type has nothing in common with the parameter's phpdoc type)
-%s:22 PhanTypeMismatchArgumentProbablyReal Argument 2 ($b) is 'A41::method' of type 'A41::method' but \test41() takes callable-array&string[] (no real type) defined at %s:8 (the inferred real argument type has nothing in common with the parameter's phpdoc type)
+%s:22 PhanTypeMismatchArgumentProbablyReal Argument 2 ($b) is 'A41::method' of type 'A41::method' but \test41() takes callable-array&non-empty-array<mixed,string> (no real type) defined at %s:8 (the inferred real argument type has nothing in common with the parameter's phpdoc type)
 %s:22 PhanTypeMismatchArgumentProbablyReal Argument 3 ($c) is (function) of type Closure():void but \test41() takes callable-string (no real type) defined at %s:8 (the inferred real argument type has nothing in common with the parameter's phpdoc type)

